core/src/test/scala/tech/beshu/ror/unit/acl/blocks/rules/GroupsAndRuleTests.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ *    This file is part of ReadonlyREST.
+ *
+ *    ReadonlyREST is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU General Public License as published by
+ *    the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    ReadonlyREST is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
+ */
+package tech.beshu.ror.unit.acl.blocks.rules
+
+import cats.data.NonEmptyList
+import eu.timepit.refined.auto._
+import org.scalatest.Inside
+import org.scalatest.matchers.should.Matchers._
+import org.scalatest.wordspec.AnyWordSpec
+import tech.beshu.ror.accesscontrol.blocks.BlockContext
+import tech.beshu.ror.accesscontrol.blocks.definitions.UserDef
+import tech.beshu.ror.accesscontrol.blocks.definitions.UserDef.Mode.WithoutGroupsMapping
+import tech.beshu.ror.accesscontrol.blocks.rules.{GroupsAndRule, GroupsRule}
+import tech.beshu.ror.accesscontrol.blocks.variables.runtime.RuntimeMultiResolvableVariable.AlreadyResolved
+import tech.beshu.ror.accesscontrol.domain.LoggedUser.DirectlyLoggedUser
+import tech.beshu.ror.accesscontrol.domain.User.Id.UserIdCaseMappingEquality
+import tech.beshu.ror.accesscontrol.domain._
+import tech.beshu.ror.unit.acl.blocks.rules.GroupsRuleTests.{RuleCreator, assertMatchRule, assertNotMatchRule, authenticationRule, groups, userIdPatterns}
```

Comment:
IMO it'd be better to refactor it a little bit and extract all the things we want to reuse to eg. a separate base test trait. In this trait, we could place all the things that are imported here.

Reply:
OK great idea. 
Now I just refactored this using an abstract class, I can make it a trait easily, but what's the difference/advantage of using a trait in this case?

Reply:
none. For me, it's ok now 👍